### ISSUE TITLE
2.7.0 -> 2.7.2 form_widget entity multiple select tag render degradation

### ISSUE THREAD

**Issue by tsifra** [symfony/symfony#15403]
After upgrading from 2.7.0 to 2.7.2 i have issue with form element rendering.
In 2.7.2 element is outputed like this (wrong way):
```html
    <select id="supplier_managers" name="supplier[managers][]" style="width:256px;" class="select multiselect" size="8" multiple="multiple">
    <option value="1">option1</option>
    <option value="13" >option2</option>
    <option value="25" >option3</option>
    <option value="4">option4</option>
    </select>
```

In 2.7.0 element is outputed like this (right way) two options with selected attribute:
```html
    <select id="supplier_managers" name="supplier[managers][]" style="width:256px;" class="select multiselect" size="8" multiple="multiple">
    <option value="1"  selected="selected">option1</option>
    <option value="13" >option2</option>
    <option value="25" >option3</option>
    <option value="4"  selected="selected">option4</option>
    </select>
```

here is entity and form definition

Entity
```php
/**
 * @var \Doctrine\Common\Collections\Collection
 *
 * @ORM\ManyToMany(targetEntity="Backend\CoreBundle\Entity\BackendUser")
 * @ORM\JoinTable(name="supplier_manager",
 *   joinColumns={
 *     @ORM\JoinColumn(name="supplier_id", referencedColumnName="supplier_id")
 *   },
 *   inverseJoinColumns={
 *     @ORM\JoinColumn(name="user_id", referencedColumnName="user_id")
 *   }
 * )
 */
private $managers;
```

Form
```php
       $builder
        ->add('managers', 'entity', array(
            'class' => 'BackendCoreBundle:BackendUser',
            'choices' => $this->options['granted_users_obj'], //object type is array BackendCoreBundle:BackendUser
            'multiple' => true,
            'choice_label' => 'printname',
            'required' => false
        ))
        ->add('btnSave', 'submit', array('label'  => 'Сохранить и продолжить'));
```

**Comment by bicpi** [comment]
I am able to reproduce this issue. It still works in Symfony 2.7.1 and then breaks in 2.7.2.

To reproduce, I used the Symfony demo app and added a Many-To-Many relation to the Post form. When selecting multiple items, the selection is gone after a failed validation in 2.7.2. This is only for a `multiple` field and when shrinking the target entities like mentioned above.

**Comment by tsifra** [comment]
If i set "expanded" property to true checkboxes are selected right way in 2.7.2. So the problem can be seen only when field is redered as selected tag

**Comment by bicpi** [comment]
@tsifra I can also confirm this behavior, it breaks only with `expanded=false`.

**Comment by bicpi** [comment]
It works again when setting strict array value comparison to `false` in `Symfony\Bridge\Twig\Extension\FormExtension::isSelectedChoice()`, i.e. entity IDs are compared as strings against the submitted IDs as integers, which obviously cannot work.

    return in_array($choice->value, $selectedValue, true);
    // changed to:
    return in_array($choice->value, $selectedValue, false);

But as this was not changed from 2.7.1 to 2.7.2, the causing changed must be somewhere else where some casting has changed. It's likely to be in #15061 I think.

**Comment by bicpi** [comment]
Removing the string type cast in `Symfony\Component\Form\ChoiceList\ArrayChoiceList:flatten()` also fixes the issue but breaks some unit tests:

    $choiceValue = (string) call_user_func($value, $choice);

But this method was introduced in #15061.

**Comment by stof** [comment]
@webmozart this string casting looks weird to me. Why is it done ?

**Comment by fbuselgarn** [comment]
This prevents us from upgrading to 2.7. Would be great if this issue gets fixed!

**Comment by tsifra** [comment]
Problem is fixed in 2.7.4. Case closed :)
